presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementMultipleDistinctAggregation.java

```diff
@@ -0,0 +1,140 @@
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.AggregationNode.Aggregation;
+import com.facebook.presto.sql.planner.plan.MarkDistinctNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.sql.planner.plan.Patterns.aggregation;
+
+/**
+ * Implements distinct aggregations with different inputs by transforming plans of the following shape:
+ * <pre>
+ * - Aggregation
+ *        GROUP BY (k),
+ *        F1(DISTINCT s0, s1, ...),
+ *        F2(DISTINCT s2, s3, ...)
+ *     - X
+ * </pre>
+ * into
+ * <pre>
+ * - Aggregation
+ *        F1(...) mask ($0)
+ *        F2(...) mask ($1)
+ *     - MarkDistinct (k, s0, s1, ...) -> $0
+ *          - MarkDistinct (k, s2, s3, ...) -> $1
+ *              - X
+ * </pre>
+ */
+public class ImplementMultipleDistinctAggregation
+        implements Rule<AggregationNode>
+{
+    private static final Pattern<AggregationNode> PATTERN = aggregation()
+            .matching(ImplementMultipleDistinctAggregation::hasDistinct);
+
+    private static boolean hasDistinct(AggregationNode aggregation)
+    {
+        return aggregation.getAggregations()
+                .values().stream()
+                .anyMatch(e -> e.getCall().isDistinct() && !e.getMask().isPresent());  // MarkDistinct doesn't support a mask (yet), so skip if aggregation has mask set
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(AggregationNode parent, Captures captures, Context context)
+    {
+        // the distinct marker for the given set of input columns
+        Map<Set<Symbol>, Symbol> markers = new HashMap<>();
+
+        Map<Symbol, Aggregation> aggregations = new HashMap<>();
+        PlanNode subPlan = parent.getSource();
+
+        for (Map.Entry<Symbol, Aggregation> entry : parent.getAggregations().entrySet()) {
+            Aggregation aggregation = entry.getValue();
+            FunctionCall call = aggregation.getCall();
+
+            if (call.isDistinct()) {
+                Set<Symbol> inputs = call.getArguments().stream()
+                        .map(Symbol::from)
+                        .collect(Collectors.toSet());
+
+                Symbol marker = markers.get(inputs);
+                if (marker == null) {
+                    marker = context.getSymbolAllocator().newSymbol(inputs.iterator().next().getName(), BOOLEAN, "distinct");
```

Comment:
just use: https://google.github.io/guava/releases/21.0/api/docs/com/google/common/collect/Iterables.html#getLast-java.lang.Iterable- instead of `inputs.iterator().next().getName()`
?